src/main/scala/chiseltest/legacy/backends/verilator/VerilatorExecutive.scala

```diff
@@ -38,16 +45,101 @@ object VerilatorExecutive extends BackendExecutive {
     // Force a cleanup: long SBT runs tend to fail with memory issues
     System.gc()
 
-    val targetDir = annotationSeq.collectFirst {
-      case firrtl.options.TargetDirAnnotation(t) => t
-    }.get
-    val targetDirFile = new File(targetDir)
+    var targetDir = annotationSeq.collectFirst {
+        case firrtl.options.TargetDirAnnotation(t) => t
+      }.get
+    var targetDirFile = new File(targetDir)
+
+    val uncachedTargetDir = targetDir
 
     val generatorAnnotation = chisel3.stage.ChiselGeneratorAnnotation(dutGen)
     val elaboratedAnno = (new chisel3.stage.phases.Elaborate).transform(annotationSeq :+ generatorAnnotation)
     val circuit = elaboratedAnno.collect { case x: ChiselCircuitAnnotation => x }.head.circuit
     val dut = getTopModule(circuit).asInstanceOf[T]
 
+    if (elaboratedAnno.contains(CachingAnnotation)) {
+      val cacheDir = s"${targetDirFile.getParent}/cache_dir"
+
+      val highFirrtlAnnos = (new ChiselStage).run(
+        elaboratedAnno :+ RunFirrtlTransformAnnotation(new HighFirrtlEmitter)
+      )
+
+      val rawFirrtl = highFirrtlAnnos
+        .collect { case EmittedFirrtlCircuitAnnotation(e) => e.value}
+        .map(l => firrtl.Parser.parse(l))
+        .head
+
+      val moduleHashString = DatatypeConverter
```

Comment:
please split this up with some more named `val`s

Reply:
Done!